File: server/review_components.py
"""Independent review fingerprints. Keep extraction identical to review-components.js."""
import re
from review_content import canonical_hash

LABELS = {'text': 'de Bijbeltekst', 'notes': 'de kanttekeningen',
          'markers': 'de nootnummers', 'citations': 'de citaatopmaak',
          'layout': 'de overige tekstopmaak', 'intro': 'de hoofdstukinleiding',
          'content': 'de inhoud'}
MARKER = re.compile(r'<sup\b[^>]*\bclass="note-marker"[^>]*>.*?</sup>', re.S)
TAG = re.compile(r'<[^>]*>')
CITATION = re.compile(r'\bclass="(?:direct-speech|[a-z]+-speaks)"')


def normalized(text):
    return re.sub(r'\s+', ' ', text or '').strip()


def position(html):
    # UTF-16 code units, also used by the browser (including supplementary characters).
    return len(normalized(TAG.sub('', html)).encode('utf-16-le')) // 2


def verse_components(verse):
    html = verse.get('text2026_html') or verse.get('text2026') or ''
    markers = [[position(MARKER.sub('', html[:match.start()])), match.group()]
               for match in MARKER.finditer(html)]
    html = MARKER.sub('', html)
    citations, layout, stack = [], [], []
    for match in TAG.finditer(html):
        tag = match.group()
        if tag.startswith('</'):
            citation = stack.pop() if stack else False
        else:
            citation = bool(CITATION.search(tag)) or (tag == '<i>' and any(stack))
            if not tag.endswith('/>'):
                stack.append(citation)
        (citations if citation else layout).append([position(html[:match.start()]), tag])
    return {
        'text': [normalized(verse.get('text2026')), normalized(TAG.sub('', html))],
        'notes': [{key: note.get(key) for key in ('marker', 'type', 'text2026')}
                  for note in (verse.get('marginNotes') or [])],
        'markers': markers, 'citations': citations, 'layout': layout,
    }


def component_payloads(kind, payload):
    if kind == 'text-verse':
        return verse_components(payload)
    if kind != 'text-chapter':
        return {'content': payload}
    verses = [(v['number'], verse_components(v)) for v in payload['verses']]
    result = {key: [[number, components[key]] for number, components in verses]
              for key in ('text', 'notes', 'markers', 'citations', 'layout')}
    result['intro'] = (payload.get('chapterIntro') or {}).get('text2026') or ''
    return result


def component_metadata(kind, payload):
    components = component_payloads(kind, payload)
    result = {key: {'revision': canonical_hash(value), 'present': (
        any(bool(item[1]) for item in value) if kind == 'text-chapter' and key != 'intro'
        else bool(value))} for key, value in components.items()}
    if kind == 'text-chapter':
        for key in result:
            if key != 'intro':
                result[key]['members'] = {str(number): canonical_hash(value) for number, value in components[key]}
    return result

File: js/review-components.js
/* Content fingerprints shared with server/review_components.py. No account data. */
(function (global) {
    'use strict';
    const labels = { text: 'de Bijbeltekst', notes: 'de kanttekeningen', markers: 'de nootnummers',
        citations: 'de citaatopmaak', layout: 'de overige tekstopmaak', intro: 'de hoofdstukinleiding', content: 'de inhoud' };
    const markerPattern = /<sup\b[^>]*\bclass="note-marker"[^>]*>.*?<\/sup>/gs;
    const tagPattern = /<[^>]*>/g;
    const normalized = text => (text || '').replace(/\s+/g, ' ').trim();
    const position = html => normalized(html.replace(tagPattern, '')).length;

    function verseComponents(verse) {
        let html = verse.text2026_html || verse.text2026 || '';
        const markers = [...html.matchAll(markerPattern)].map(match =>
            [position(html.slice(0, match.index).replace(markerPattern, '')), match[0]]);
        html = html.replace(markerPattern, '');
        const citations = [], layout = [], stack = [];
        for (const match of html.matchAll(tagPattern)) {
            const tag = match[0];
            let citation;
            if (tag.startsWith('</')) citation = stack.length ? stack.pop() : false;
            else {
                citation = /\bclass="(?:direct-speech|[a-z]+-speaks)"/.test(tag) || (tag === '<i>' && stack.some(Boolean));
                if (!tag.endsWith('/>')) stack.push(citation);
            }
            (citation ? citations : layout).push([position(html.slice(0, match.index)), tag]);
        }
        return { text: [normalized(verse.text2026), normalized(html.replace(tagPattern, ''))],
            notes: (verse.marginNotes || []).map(note => Object.fromEntries(
                ['marker', 'type', 'text2026'].map(key => [key, note[key] ?? null]))), markers, citations, layout };
    }

    function payloads(kind, payload) {
        if (kind === 'text-verse') return verseComponents(payload);
        if (kind !== 'text-chapter') return { content: payload };
        const verses = payload.verses.map(v => [v.number, verseComponents(v)]);
        const result = Object.fromEntries(['text', 'notes', 'markers', 'citations', 'layout'].map(key =>
            [key, verses.map(([number, components]) => [number, components[key]])]));
        result.intro = payload.chapterIntro?.text2026 || '';
        return result;
    }

    function canonical(value) {
        if (Array.isArray(value)) return '[' + value.map(canonical).join(',') + ']';
        if (value && typeof value === 'object') return '{' + Object.keys(value).sort()
            .map(key => JSON.stringify(key) + ':' + canonical(value[key])).join(',') + '}';
        return JSON.stringify(value);
    }

    async function revisions(kind, payload) {
        return Object.fromEntries(await Promise.all(Object.entries(payloads(kind, payload)).map(async ([key, value]) => {
            const bytes = new TextEncoder().encode(canonical(value));
            const hash = await crypto.subtle.digest('SHA-256', bytes);
            return [key, [...new Uint8Array(hash)].map(n => n.toString(16).padStart(2, '0')).join('')];
        })));
    }

    function join(keys) {
        return new Intl.ListFormat('nl', { style: 'long', type: 'conjunction' }).format(keys.map(key => labels[key]));
    }

    function warning(parts) {
        const groups = { changed: [], pending: [], correction: [], stale: [], local: [] };
        for (const [key, part] of Object.entries(parts)) {
            if (part.local) groups.local.push(key);
            else if (part.stale) groups.stale.push(key);
            else if (part.status === 'correction-needed') groups.correction.push(key);
            else if (part.status !== 'approved') groups[part.needsReverification ? 'changed' : 'pending'].push(key);
        }
        const sentences = [];
        if (groups.local.length) sentences.push('lokale bewerkingen aan ' + join(groups.local) + ' zijn niet geverifieerd.');
        if (groups.changed.length) sentences.push(join(groups.changed) +
            (groups.changed.length === 1 && !['notes', 'markers'].includes(groups.changed[0]) ? ' is' : ' zijn') +
            ' gewijzigd sinds de laatste verificatie en nog niet opnieuw gecontroleerd.');
        if (groups.pending.length) sentences.push(join(groups.pending) +
            (groups.pending.length === 1 && !['notes', 'markers'].includes(groups.pending[0]) ? ' is' : ' zijn') +
            ' nog niet geverifieerd.');
        if (groups.correction.length) sentences.push('voor ' + join(groups.correction) + ' is een aanpassing aangevraagd.');
        if (groups.stale.length) sentences.push('de getoonde versie van ' + join(groups.stale) +
            ' wijkt af van de actuele versie. Herlaad om de verificatiestatus te controleren.');
        return sentences.length ? 'Let op: ' + sentences.join(' ') : '';
    }

    const correctionLabel = task => task.component === 'custom' ? task.customTarget : labels[task.component];
    global.ReviewComponents = { labels, payloads, revisions, join, warning, correctionLabel };
})(typeof window === 'undefined' ? globalThis : window);
